### DISCUSSION TITLE
Setting registry not being reflected from nx executor options

### DISCUSSION THREAD

**Discussion by ozee** [bikecoders/ngx-deploy-npm#519] [Q&A] [answered]
I am trying to use the `registry` option to define our private npm repository but it seems like no matter what I do, it doesn't recognize that the option is being set. I have also tried setting it via `$HOME/.npmrc` prior to running the nx deploy script both as a scoped registry and a catch-all. I also have my publishConfig set in my package.json that is in the dist folder being deployed but I think that is also ignored or being overwritten. The strange part is that it is getting set somewhere because our proxy the npm registry is being set, but I am unable to publish to that, we have an internal url for publishing to. Here is the relevant nx target:

```json
"deploy": {
  "executor": "ngx-deploy-npm:deploy",
  "options": {
    "access": "public",
    "dryRun": true,
    "noBuild": true,
    "distFolderPath": "dist/packages/somePackage/",
    "registry": "https://****-internal-npm/"
  }
},
```

Any help would be greatly appreciated!

**Comment by dianjuar** [comment]
Identifying misconfiguration could be tricky. It depends so much on the context you are trying to do it, for instance, are you trying it in a local machine or ci?

To try to identify the problem, it would be tremendously helpful to provide as much information as you can. This template could help.

```markdown
### Current Behavior
<!-- What is the behavior that currently you experience? -->

### Expected Behavior
<!-- What is the behavior that you expect to happen? -->

### GitHub Repo
<!-- This is extremely important! If this issue is reproduce-able on https://github.com/nrwl/nx-examples, you may use that as the repo.
If not, please do take a few minutes of your time to create a repo to help us reproduce the issue.
This is the best way to help us reproduce the issue and fix it as soon as possible. -->

### Steps to Reproduce
<!-- Please provide some instructions to reproduce the issue in the repo provided above. Be as detailed as possible. -->

1.

### Nx Report
<!-- Please paste the contents shown by nx report. This will be automatically formatted into code, so no need for backticks. -->

### Failure Logs
<!--  Please include any relevant log snippets or files here  -->

### Additional Information
<!-- Is there any additional information that you can provide? -->
```

Pasting the command `nx deploy YOUR_LIB` output will be helpful. Setting the npm registry can be done in so many places that output will give us a clue of which configuration is taking place.

This might help to identify why it isn't working. There are some of the places that are at the top of my head:
- Env Variables
  - Sometimes, your CI does that for you
- `$HOME/.npmrc`
- `PROJECT_HOME/.npmrc`
- I think `LIB_HOME/.npmrc`
- Library `package.json`

---

It would be interesting to identify if something is already set up that npm is overwriting your instructions. For instance, env variables had the ultimate priority (I believe so).

An info you might find helpful is that the option `registry` is the same as doing `npm --registry`. If something else has priority over the option `--registry` on the command line, it must be removed if you desire the option registry prevails.

**Comment by ozee** [comment]
Thanks for the quick response @dianjuar! We actually use that same template internally, haha, I should have filled out the information for throughly.

### Current Behavior

The registry isn't set whenever you define it in options.

### Expected Behavior

That whenever you define the registry in the option that it sets that registry, overwriting other places where it could have pulled from.

### GitHub Repo

I don't have a public repo currently that I can show, however I think that you may be onto something above.

### Steps to Reproduce

Try to publish and the deploy packages, gets another registry location.

### Nx Report

```sh
   Node : 16.19.1
   OS   : darwin arm64
   npm  : 8.19.3
   
   nx                      : 15.8.6
   @nrwl/js                : 15.8.6
   @nrwl/jest              : 15.8.6
   @nrwl/linter            : 15.8.6
   @nrwl/workspace         : 15.8.6
   @nrwl/cli               : 15.8.6
   @nrwl/cypress           : 15.8.6
   @nrwl/devkit            : 15.8.6
   @nrwl/eslint-plugin-nx  : 15.8.6
   @nrwl/nx-plugin         : 15.8.6
   @nrwl/storybook         : 15.8.6
   @nrwl/tao               : 15.8.6
   @nrwl/vite              : 15.8.6
   @nrwl/nx-cloud          : 15.3.5
   typescript              : 4.9.5
   ---------------------------------------
   Community plugins:
   @jscutlery/semver       : 2.30.1
   @nxkit/style-dictionary : 2.3.0
   ngx-deploy-npm          : 5.2.0
```

### Additional Information

The output from our command `nx deploy YOUR_LIB` is success until literally the final step of the publish where it cites a different registry from what I tried to define.

Some background, we do set the registry earlier in our action to our public npm-proxy for running `npm i` within our workflow inside our organization. I think that locally it may be referring to my `$HOME/.npmrc` directory when running a dry-run. I think that I will try adding a `.npmrc` to each component to see if that does anything different.

**Comment by dianjuar** [comment]
Can we see the output of the `npx nx deploy YOUR_LIB --dry-run` ?

**Comment by ozee** [comment]
```sh
 *  Executing task: npx nx run ****:deploy 


> nx run ****:deploy

📦 Building "****"
📦 Build target "****"
vite v4.3.3 building for production...
✓ 1 modules transformed.
../../../dist/packages/****/controllers/theme-context/index.js  2.63 kB │ gzip: 0.88 kB │ map: 8.96 kB
[vite-plugin-static-copy] Copied 2 items.
✓ built in 64ms
Dry-run: The package is not going to be published
npm
 notice 
npm notice 📦  ****@0.0.2-next.0
npm notice === Tarball Contents === 

npm notice 3.6kB README.md   
npm
 notice 814B  package.json
npm notice === Tarball Details === 

npm notice
 name:          ****                       
npm notice version:       0.0.2-next.0                            
npm notice filename:      ****-0.0.2-next.0.tgz      
npm notice package size:  1.9 kB                                  
npm notice unpacked size: 4.4 kB                                  
npm notice shasum:        54deec2c7947473ea7abad6f98c7c0a6f60c93d1
npm notice integrity:     sha512-JlRYiBqkD+/Ae[...]7o1i9ZZpH+GqQ==
npm notice total files:   2                                       
npm notice 

npm WARN This command requires you to be logged in to https://****/artifactory/api/npm/npm-public/ (dry-run)

npm notice Publishing to https://****/artifactory/api/npm/npm-public/ (dry-run)

+ ****@0.0.2-next.0

The options are:
{
 "access": "public",
 "registry": "https://****/artifactory/api/npm/****-internal-npm/",
 "dryRun": true,
 "noBuild": false
}
🚀 Successfully published via ngx-deploy-npm! Have a nice day!
```

I forgot to add my dist target directory, but that's not really needed to showcase what happened. On my local machine, I know that I have the npm-public proxy set in my $HOME/.npmrc, so after reading what you wrote that made a lot of sense.

**Comment by dianjuar** [comment]
The output provided seems to work correctly in the environment you tested (the local machine, I believe).

Also, I understand that you always want those libs to publish in your private repo. If that is the case, the `--registry` is unadvised. The `--registry` option is made to overwrite what is already established. Useful for those cases when you say _"I want to do an exception this time or on this particular scenario"_, not to make it a rule.

I'd like to encourage you to make sure you have the proper configuration to, as a default, publish the libraries on your private repo. That can be done by modifying your lib's `package.json`.

I was hoping you could take a look at the npm documentation about [Preventing my package from being published on the official registry](https://docs.npmjs.com/cli/v8/using-npm/registry#how-can-i-prevent-my-package-from-being-published-in-the-official-registry)

**Comment by ozee** [comment, marked as answer]
I found the culprit last yesterday, I had a `.npmrc` file in my project's root that was pointing our scoped project always to a specific place. So no matter how much I created or changed it at the user level it was always being overridden by it. I appreciate you taking the time to look over this with me!
